python/cudf/cudf/io/parquet.py

```diff
@@ -17,21 +18,59 @@
 from cudf.utils import ioutils
 
 
-def _get_partition_groups(df, partition_cols, preserve_index=False):
-    # TODO: We can use groupby functionality here after cudf#4346.
-    #       Longer term, we want more slicing logic to be pushed down
-    #       into cpp.  For example, it would be best to pass libcudf
-    #       a single sorted table with group offsets).
-    df = df.sort_values(partition_cols)
-    if not preserve_index:
-        df = df.reset_index(drop=True)
-    divisions = df[partition_cols].drop_duplicates(ignore_index=True)
-    splits = df[partition_cols].searchsorted(divisions, side="left")
-    splits = splits.tolist() + [len(df[partition_cols])]
-    return [
-        df.iloc[splits[i] : splits[i + 1]].copy(deep=False)
-        for i in range(0, len(splits) - 1)
+def _write_parquet(
+    df,
+    paths,
+    compression="snappy",
+    index=None,
+    statistics="ROWGROUP",
+    metadata_file_path=None,
+    int96_timestamps=False,
+    row_group_size_bytes=None,
+    row_group_size_rows=None,
+    partitions_info=None,
+    **kwargs,
+):
+    if is_list_like(paths) and len(paths) > 1:
+        if partitions_info is None:
+            ValueError("partition info is required for multiple paths")
+        elif not is_list_like(partitions_info):
+            ValueError("partition info must be list-like for multiple paths")
+        elif not len(paths) == len(partitions_info):
+            ValueError("partitions_info and paths must be of same size")
+    if is_list_like(partitions_info) and len(partitions_info) > 1:
+        if not is_list_like(paths):
+            ValueError("paths must be list-like when partitions_info provided")
+
+    paths_or_bufs = [
+        ioutils.get_writer_filepath_or_buffer(path, mode="wb", **kwargs)
+        for path in paths
     ]
+    common_args = {
+        "index": index,
+        "compression": compression,
+        "statistics": statistics,
+        "metadata_file_path": metadata_file_path,
+        "int96_timestamps": int96_timestamps,
+        "row_group_size_bytes": row_group_size_bytes,
+        "row_group_size_rows": row_group_size_rows,
+        "partitions_info": partitions_info,
+    }
+    if all([ioutils.is_fsspec_open_file(buf) for buf in paths_or_bufs]):
+        with ExitStack() as stack:
+            fsspec_objs = [stack.enter_context(file) for file in paths_or_bufs]
+            file_objs = [
+                ioutils.get_IOBase_writer(file_obj) for file_obj in fsspec_objs
+            ]
+            write_parquet_res = libparquet.write_parquet(
+                df, filepaths_or_buffers=file_objs, **common_args
+            )
```

Comment:
Python reviewers, is this an appropriate replacement for https://github.com/rapidsai/cudf/blob/ba3aedbb869ef95cae517f56ffb2bab305bffa40/python/cudf/cudf/io/parquet.py#L702-L708

Note that because we want to open multiple files to write to them at the same time, I couldn't use `with`.

Reply:
I think this is reasonable.